Fix for Amazon purchase dialog not showing up

vegaro opened the pull request:
[CF-786]


purchases-android has to be initialized on application onCreate otherwise the purchasing dialog for Amazon doesn’t show up in production.

The reason why this happens is because when initiating a purchase, the Amazon app store checks if there’s a UI to show the dialog on top of. We know this by looking at the logs. When an Activity starts, there’s a log indicating that the onCreate of the Activity is called, same for the onDestroy . They probably have an implementation similar to what’s suggested in https://medium.com/the-devops-corner/how-to-detect-android-app-foreground-status-c9443ddef260

We used to have an implementation similar to that one in our SDK to determine if the app was on foreground. We removed it and started using androidx Lifecycle because that implementation has a big flaw: it only works if the Activity lifecycle listener is added when the Application onCreate is called. It relies on listening to the onCreate of an Activity to be called to determine if the app is in foreground. If the Activity lifecycle listener is added after the Activity onCreate is called, it will not detect the app is in foreground (since onCreate has already be called).

If developers call configure (which is where we set the Amazon listener and where Amazon will start listening to Activity lifecycles) outside of Application’s onCreate and after an Activity onCreate has been called, Amazon will think the app is not in foreground and will not show the purchase dialog in production. For some reason it works when the app is executed in debug mode. I think the reason is that the code to check if the app is in foreground lives in the Amazon Appstore, and Amazon App Tester doesn’t perform the check, because I don’t see any logs.

One solution, is to be very clear that the SDK has to be configured in Application’s onCreate. The problem is that in hybrids developers don’t have easy access to the Application onCreate.

Another solution would be to open a transparent Activity when performing the purchase. This is basically what Google does with ProxyBillingActivity https://github.com/DimaDake/billing/blob/ea2fe60d54832dc8890bd4857841808b13afb9e5/library/src/main/java/com/android/billingclient/api/ProxyBillingActivity.java

tonidero commented on feature/amazon/src/main/java/com/revenuecat/purchases/amazon/ProxyAmazonBillingActivity.kt:
```diff
@@ -0,0 +1,23 @@
+        }
+        resultReceiver?.send(0, bundle)
+
+        finish()
```
Hmm we shouldn't need to do this I think? We will finish the activity from the `DefaultPurchasingServiceProvider` once `onPurchaseCompleted` is called. As in, I think right now we would be trying to finish this activity twice

vegaro replied:
Oh yes, this is leftovers

tonidero commented on feature/amazon/src/main/java/com/revenuecat/purchases/amazon/AmazonBilling.kt:
```diff
@@ -232,13 +232,19 @@ internal class AmazonBilling constructor(
                     storeProduct,
                     presentedOfferingIdentifier,
+                    onPurchaseCompleted = {
+                        purchaseHandler.onPurchaseCompleted(activity)
```
This will hold a reference to the `activity` until this is done... There could be some issues if this never ended for any reason. One hacky possibility is to pass a `WeakReference` to the activity... Not very clean though...

aboedo replied:
genuine question as an iOS dev used to weak references: what's the not-clean part?

tonidero replied:
Well, in Java/Kotlin, it's not common to hold weak references since it's the norm to rely on the GC to free memory. I would say it's a technique only used when absolutely necessary and it's generally a better idea to rearchitect things as needed to make sure we don't hold anything that shouldn't be stored. I don't think there is a technical reason, and it's more of an architecture/style thing.

That said, in this case, it could cause actual issues since it's possible for the original activity to be freed and we never `finish` the `ProxyAmazonBillingActivity` so it could cause us to leave that orphaned activity as well.

vegaro commented on feature/amazon/src/main/java/com/revenuecat/purchases/amazon/handler/PurchaseHandler.kt:
```diff
@@ -24,30 +29,50 @@ class PurchaseHandler(
+                synchronized(this@PurchaseHandler) {
+                    val requestId = resultData?.get("request_id") as? RequestId
+                    if (requestId != null) {
+                        purchaseCompletedCallbacks[requestId] = onPurchaseCompleted
```
That's true... We actually error out before getting to this point indicating there's another purchase in progress

vegaro commented:
@tonidero I like the BroadcastReceiver idea. We have the application context. We use it for registering the Amazon listener in AmazonBilling, and we actually hold it in AmazonBilling.

aboedo commented on feature/amazon/src/main/java/com/revenuecat/purchases/amazon/DefaultPurchasingServiceProvider.kt:
```diff
@@ -1,11 +1,15 @@
 import com.amazon.device.iap.model.FulfillmentResult
 import com.amazon.device.iap.model.RequestId
 
+const val PROXY_AMAZON_BILLING_ACTIVITY_REQUEST_CODE = 192342
```
what in the world

aboedo replied:
how did you determine this number?

tonidero replied:
That's basically a "tag" that gets set when starting an activity and allows to determine the result corresponds for this specific code, so it doesn't really matter what value it has. However, I believe it has to be a number less than `65535`, so I actually think this number is not valid. That said, if we go with the `BroadcastReceiver` idea, we won't need to pass this value, so it can be removed then

tonidero commented:
I was thinking about something we should test with this approach of starting a custom activity while the purchase process is ongoing. That is, make sure the theme of the original activity is respected, or at least that it doesn't do anything visually funky. For example, if the original activity is in full screen mode, it's possible we leave full screen when we start this proxy activity, and depending how the original activity is setup, it's possible we don't return to full screen after the purchase is completed/cancelled/errored (for example if we go to full screen on the original activity `onCreate` which won't get called when going back to that activity or on a specific part of that activity which doesn't get called in the `onResume`). 

I'm thinking when we start the new activity, we might have to pass some flags and the theme to the new activity so we can follow the same looks, but then again I think we should test this to make sure it's a problem.

vegaro commented:
Leaving this here https://developer.android.com/topic/libraries/app-startup. I think with this we can hook to app startup using that Google library. Then we could register the Amazon listener when the app initializes, which will make sure the Amazon listener is registered before any Activity is created. The problem is that involves SO MUCH refactor since in order to register the listener we need to have initialized the listener first which right now is `AmazonBilling` and has so many dependencies

tonidero commented:
Good find! I hadn't thought about that one. I didn't know it allows libraries to configure startup things as well.

>  we don't know if the developers want to configure for the Amazon App store until they call configure

We could kind of assume they do if they include the amazon sdk dependency I guess? It's not great to do those assumptions though, so not sure if we would have to add some config files to allow to disable it... (again, not a great option). 

Having said that, I haven't used any libraries that use that google library afaik, so not sure how well that would work. I know starting a proxy activity as we were planning to do can potentially cause some issues in specific situations, so I don't like that option much either. If I had to vote for one approach, I feel using this library would be less hacky, but I'm not familiar with how much refactor we would have to do to make it work. If it's too much as you say, it might make sense to try with the proxy activity approach first, since we were close and see if we indeed have any issues with the theming and/or full screen modes.

vegaro commented:
Regarding the theming issues... I am not extremely concerned because Google Billing Library has an ProxyBillingActivity and I copied their manifest. But we should definitely test it thoroughly. I might need some help with that.

vegaro commented:
I am actually seeing the notification bar going from top to bottom of screen when the activity is terminated 👎

tonidero commented:
Hmm you mean the status bar right? Could you record a short video to see how it looks? We might be able to fix it with some flags. And yeah not sure if Google's proxy activity sets up these flags as well... They are usually set in the manifest or in the activity itself... And I'm happy to help with testing this, we probably can try to play with the theme in the activity in one of our example apps

vegaro commented:
@tonidero This issue is basically what I am seeing https://github.com/android/play-billing-samples/issues/135

Although I am trying the solution and now I see a black background 🤔 Also... those theme flags are only available for Android 21+ and our min sdk is 14 (Amazon's is also 14)

vegaro commented:
It could definitely introduce incompatibilities. By the way, we managed to fix the status bar animation yesterday, huge props to @tonidero for helping me out. I am going to be testing it today in a Fire Tv to make sure it looks ok and in an Android device with Android 7.0

tonidero commented on feature/amazon/src/test/java/com/revenuecat/purchases/amazon/handler/PurchaseHandlerTest.kt:
```diff
@@ -194,6 +271,15 @@ class PurchaseHandlerTest {
         assertThat(receivedCount).isOne
     }
 
+    private fun verifyActivityIsStartedAndFakeRequestId(dummyRequestId: String) {
```
Personal preference but I usually prefer private methods to go after the tests

vegaro commented on feature/amazon/src/main/AndroidManifest.xml:
```diff
@@ -3,13 +3,19 @@
+                android:name=".ProxyAmazonBillingActivity"
+                android:configChanges="keyboard|keyboardHidden|screenLayout|screenSize|orientation"
+                android:exported="false"
+                android:theme="@android:style/Theme.Translucent.NoTitleBar" />
```
It's redundant but I added it as a backup in case `setTheme` in the activity doesn't work. Looking at all the problems we had setting the theme the other day I wouldn't be surprised `setTheme` doesn't work. What do you think? I am happy to remove it too

tonidero replied:
Well, I think it would be confusing to have it in both places... Unless we know this is needed I think I would vote to remove it. Not a strong opinion though.

tonidero commented on feature/amazon/src/main/java/com/revenuecat/purchases/amazon/ProxyAmazonBillingActivity.kt:
```diff
@@ -0,0 +1,43 @@
+            val sku = intent.getStringExtra("sku")
+            val resultReceiver = intent.getParcelableExtra<ResultReceiver>("result_receiver")
+            val purchasingServiceProvider =
+                intent.getParcelableExtra<PurchasingServiceProvider>("service_provider")
```
Another idea could be to create a `ProxyAmazonBillingActivityArgs` class that contains all these 3 and make it parcelable, so there is only one argument. That way is slightly safer since the arguments definition is also centralized

vegaro commented on feature/amazon/src/main/AndroidManifest.xml:
```diff
@@ -3,13 +3,19 @@
+        <activity
+                android:name=".ProxyAmazonBillingActivity"
+                android:configChanges="keyboard|keyboardHidden|screenLayout|screenSize|orientation"
+                android:exported="false"
```
`exported` needs to be explicitely set when targeting Android 12 apps, otherwise the installation fails https://developer.android.com/about/versions/12/behavior-changes-12#exported

vegaro commented on feature/amazon/src/main/java/com/revenuecat/purchases/amazon/ProxyAmazonBillingActivityBroadcastReceiver.kt:
```diff
@@ -0,0 +1,16 @@
+import android.content.Context
+import android.content.Intent
+
+internal class ProxyAmazonBillingActivityBroadcastReceiver(private val activity: Activity) : BroadcastReceiver() {
```
Yes, I don't like this... Do you think it would be a better idea to make `ProxyAmazonBillingActivity` implement `BroadcastReceiver`? That way it will finish itself

vegaro replied:
ugh yes, it's a class, I didn't notice. Hmmmm. Yeah I don't think there's any alternative. I am calling unregister in the `onDestroy` and setting it to null so I don't think it will leak, but who knows.

vegaro replied:
Done in [`2cbd49c` (#552)](https://github.com/RevenueCat/purchases-android/pull/552/commits/2cbd49c1c0bc6f40094c75f5e8896fbf7aa2fd5e)

vegaro commented on feature/amazon/src/test/java/com/revenuecat/purchases/amazon/ProxyAmazonBillingActivityBroadcastReceiverTest.kt:
```diff
@@ -0,0 +1,53 @@
+    fun `ProxyAmazonBillingActivityBroadcastReceiver sets onReceiveCalled`() {
+        assertThat(underTest.onReceiveCalled).isFalse
+        underTest.onReceive(null, null)
+        assertThat(underTest.onReceiveCalled).isTrue
```
That's the other test. I might just remove this test

vegaro commented on feature/amazon/src/main/java/com/revenuecat/purchases/amazon/handler/PurchaseHandler.kt:
```diff
@@ -47,7 +88,13 @@ class PurchaseHandler(
+            Intent().also { intent ->
+                intent.action = "purchase_finished"
+                intent.setPackage(applicationContext.packageName)
+                applicationContext.sendBroadcast(intent)
```
Would it make sense to move it to `ProxyAmazonBillingActivityBroadcastReceiver` as a static function? Or keep it in PurchaseHandler. I am not sure. I am leaning towards keeping it in PurchaseHandler, but I would like to keep all this Broadcast logic in the same place

vegaro commented:
@tonidero I think I've finished addressing your comments. I will fix the tests after your review, in case you think other changes to the architecture should be made.

tonidero commented on feature/amazon/src/main/java/com/revenuecat/purchases/amazon/ProxyAmazonBillingActivityBroadcastReceiver.kt:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    var onReceiveCalled = false
```
So this is only added for a test... I'm not sure but is there a way of testing whether the activity was finished in the test scenario? If so we could remove this property entirely. If not, I guess we can keep it for better tests.

vegaro replied:
I tried but I couldn't find anything 😢 I thought about checking if state is `ON_DESTROY` but it wasn't passing because `sendBroadcast` is asynchronous. Maybe I can mock that?

tonidero replied:
It will probably be hard to mock, since it's created by the activity. I think it's ok, we can leave as is 👍

tonidero commented on feature/amazon/src/main/java/com/revenuecat/purchases/amazon/ProxyAmazonBillingActivityBroadcastReceiver.kt:
```diff
@@ -0,0 +1,38 @@
+        activity.get()?.finish()
+    }
+
+    companion object {
```
:nitpick: I usually prefer having companion objects at the top of the file.

tonidero commented on feature/amazon/src/main/java/com/revenuecat/purchases/amazon/ProxyAmazonBillingActivityBroadcastReceiver.kt:
```diff
@@ -0,0 +1,38 @@
+
+        fun newPurchaseFinishedIntent(applicationContext: Context): Intent {
+            return Intent(PURCHASE_FINISHED_ACTION).also { intent ->
+                intent.setPackage(applicationContext.packageName)
```
Since this requires only the package name, I would probably pass that as parameter of the `newPurchaseFinishedIntent` instead of the entire context.

tonidero reviewed (commented):
It's looking pretty good! Just some small things. Lmk once it's final and I can give it a final review.

vegaro commented on feature/amazon/src/main/java/com/revenuecat/purchases/amazon/ProxyAmazonBillingActivity.kt:
```diff
@@ -0,0 +1,51 @@
+        super.onCreate(savedInstanceState)
+
+        proxyAmazonBillingDelegate = ProxyAmazonBillingDelegate()
+        proxyAmazonBillingDelegate?.onCreate(this, savedInstanceState)
```
@tonidero any ideas on how to test that this is called? I could do some little DI framework that injects a mocked `ProxyAmazonBillingDelegate` but I think that's too much lol

tonidero commented on feature/amazon/src/main/java/com/revenuecat/purchases/amazon/purchasing/ProxyAmazonBillingDelegate.kt:
```diff
@@ -0,0 +1,64 @@
+
+    fun onCreate(activity: Activity, savedInstanceState: Bundle?) {
+        broadcastReceiver = ProxyAmazonBillingActivityBroadcastReceiver(activity)
+        activity.applicationContext.registerReceiver(broadcastReceiver, filter)
```
Hmm I noticed we are now registering the receiver in the application context, instead of the `ProxyAmazonBillingActivity`. I think it would be safer on the activity, since if we leave it on the application, and the `onDestroy` doesn't get called for any reason, the application can remain registered forever.

tonidero commented on feature/amazon/src/main/java/com/revenuecat/purchases/amazon/purchasing/ProxyAmazonBillingDelegate.kt:
```diff
@@ -0,0 +1,64 @@
+        val purchasingServiceProvider =
+            intent.getParcelableExtra<PurchasingServiceProvider>(
+                ProxyAmazonBillingActivity.EXTRAS_PURCHASING_SERVICE_PROVIDER
+            )
```
I was wondering whether getting the arguments from the intent could be done in the activity... But I guess it's fine here so we have that logic available for testing here as well 👍.

tonidero commented on feature/amazon/src/test/java/com/revenuecat/purchases/amazon/ProxyAmazonBillingActivityBroadcastReceiverTest.kt:
```diff
@@ -0,0 +1,47 @@
+
+    @After
+    fun tearDown() {
+        clearAllMocks()
```
NABD but currently we are reseting all mocks on the `@Before` block, so this isn't really needed. If we created the mocks and didn't create new ones for each test, then this would be needed.

vegaro replied:
Oh yes, true, I will remove it

tonidero reviewed (approved):
I think this looks good!

vegaro commented:
@tonidero I removed the clearAllMocks and moved the ProxyBillingActivity stuff to a `purchasing` package
